docs/design/kademlia-removal.md

```diff
@@ -0,0 +1,84 @@
+# Kademlia Removal
+
+## Abstract
+
+This design document outlines the communication protocol between satellites and 
+storage nodes, network refreshes, and kademlia removal given the satellite opt-in 
+capability for storage nodes. 
+
+## Background
+User research on how Storage Node Operators would connect to satellites resulted in a 
+simple solution called Opt-in SNO select. Satellites wait for storage nodes to work 
+with them, and storage nodes manually select the satellites they want to work with. 
+The initial implementation of this allows SNOs to update their trusted satellite list 
+in a hardcoded configuration file, but future improvements will enable users to manage 
+this list through a web console. Due to the nature of this change, the Storj network 
+at this point in time no longer requires the Kademlia DHT or related entities. 
+We will replace this protocol with direct communication between satellites and 
+storage nodes, and keep the network fresh without kademlia node discovery and random 
+lookups.
+
+## Design
+
+1. Nodes reach out to satellites that they want to work with
+    
+    - At the time of writing, a storage node operator can input a list of trusted satellite 
+    IDs and addresses into their configuration file on setup. There are several tardigrade-level 
+    satellites already listed as a preset. This list is imported into a Go map data structure 
+    within the “trust” package.
+    - Eventually during the design and implementation of trusted satellite management, 
+    this configured list may be imported into a on-disk data store, eg SQL table. Users will 
+    be able to modify this list through the storage node operator console. However, this is a 
+    separate discussion from this design document. This design is satisfied with a static trusted 
+    satellites list.
+    - Nodes will be able to communicate with satellites through the “transport” and “trust” packages 
+    directly rather than using kademlia to traverse the network to find the address of a given ID.
+
+2. Network refreshes at a regular interval
+    - Nodes will keep themselves up to date in the network by pinging all the satellites in their 
+    trusted list every hour. 
+    - Satellites will ping the nodes back
+        - If is it successful, the satellite will insert or update the node in the overlay cache and 
+        notify the node of success
+        - If not, it does not proceed with updating the overlay cache, and the node receives an error message
+    - This will use grpc for now, but future development may use a non-ssl connection from node to satellite 
```

Comment:
SSL also has a significant resource and performance impact. any place we can avoid SSL will make us faster. in particular, this specific place is safe to not have over SSL. it's basically an anonymous tip box for satellites to find out about new nodes. satellites must still validate that they can talk to the node over ssl, but the tip box part does not need to be over ssl

Reply:
> it's basically an anonymous tip box for satellites to find out about new nodes
Can I claim that you IP has changed? Without ssl I don't need a private key to send that message. The ssl connection is preventing this very good.

> satellites must still validate that they can talk to the node over ssl
Why do we need 2 steps for that? Can't we add/update and validate the information in one step?

Reply:
In this scenario, is the SN going to use a non ssl connection for every uptime ping or just the initial contact?

Reply:
> Can I claim that you IP has changed? Without ssl I don't need a private key to send that message. The ssl connection is preventing this very good.

Sure, but the satellite won't trust that until it connects to the storage node to validate.

> Why do we need 2 connections for that? Can't we add/update and validate the information with one ssl connection?

We need two connections because the Satellite needs to confirm that the given ip/port will actually reach the node in question. The IP/port provided could be bad. We need to test with an *outgoing* connection from the satellite.

> In this scenario, is the SN going to use a non ssl connection for every uptime ping or just the initial contact?

 just the initial suggestion to the satellite of where to find the storage node

Reply:
How about a man in the middle attack? I tell the satellite that the ip and port for your node has changed. The satellite will try to contact me and I will redirect that to your real ip and port.
The connection is encrypted? Which keypair is used for the encryption? Can I get a copy of that keypair and read all the packages? Even if that is not possible I could wait for the initial handshake and than take your node offline.

Reply:
Okay, you're right, the UDP messages need to be signed by the node's identity. Good catch.

Reply:
What about a DDOS attack here (or something similar -- I don't know all the formal names for attack vectors)? For instance, if I maliciously drop 20 million "tips" into the tip-box, the satellite now has too much work to do, it'll never get around to pinging real nodes.

It seems you should only be able to request a ping-back for yourself, and that rate-limiting it is a good idea.

Reply:
that's a reason why they should be signed. there's not much more we can do. flooding the tip box with signed udp packets is pretty much the same as DoSing the satellite with ssl handshakes. but yeah we should make sure they're signed so we can at least throw out nonsense requests

Reply:
Signed with what? If we have a signed UDP packet for the very first ping from a node, what do we use to verify the signature?

Reply:
ooohh hot dang

so okay i guess the packet needs to include
1) the address of the node
2) the certificate chain to be expected once talking to the node that identifies the node
3) a signature of the above things with the leaf private key of that certificate chain.

this does mean you can just generate certificates willy nilly and submit whatever address you want to the tip box, but they will be ignored if
1) the difficulty of the computed id isn't high enough
2) the node you end up talking to doesn't have the same node id as the one computed from the tipster certificate chain.

Reply:
> ooohh hot dang

Making you say that is what I live for.

> they will be ignored

Ignored in the long run, but only after the satellite has done a significant amount of work (ie pinging the node to verify everything).

That said, I think the POW in the nodeID saves us here. If we (1) de-dupe so that we're not processing the same nodeID over and over, (2) check the POW first thing, then I _think_ it becomes not worth it for a troll to do this just for fun.

Reply:
> Making you say that is what I live for.

lol oh no, we should get you some additional life goals

but yeah, great catch

Reply:
Would we like to create this UDP tip box in this iteration, or save it for future improvements?

Reply:
Ok I decided that we will use UDP in a future improvement